ibis/expr/tests/test_signature.py

```diff
@@ -0,0 +1,234 @@
+    assert isinstance(Between.signature, TypeSignature)
+    assert Between.signature.names() == argnames
+    assert Between.__slots__ == argnames
+
+    obj = Between(10, lower=2)
+    assert obj.value == 10
+    assert obj.lower == 2
+    assert obj.upper is None
+
+    assert obj.args == (10, 2, None)
+    assert obj.argnames == argnames
+
+
+def test_maintain_definition_order():
+    class Between(Annotable):
+        value = Argument(int)
+        lower = Argument(int, default=0)
+        upper = Argument(int, default=None)
+
+    assert list(Between.signature.keys()) == ['value', 'lower', 'upper']
+
+    if PY2:
+        assert (Between.signature['value']._serial <
+                Between.signature['lower']._serial <
+                Between.signature['upper']._serial)
+
+
+def test_signature_equals():
+    s1 = TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(int))
+    ])
+    s2 = TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(int))
+    ])
+    s3 = TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(float))
+    ])
+    s4 = TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(float))
+    ])
+    s5 = TypeSignature([
+        ('left_one', Argument(int)),
+        ('right', Argument(float))
+    ])
+    s6 = TypeSignature([
+        ('left_one', Argument(int)),
+        ('right', Argument(int))
+    ])
+    assert s1 == s2
+    assert s3 == s4
+    assert s1 != s3
+    assert s2 != s4
+    assert s1 != s5
+    assert s2 != s6
+    assert s5 != s6
+
+
+def test_signature_inheritance():
+    class IntBinop(Annotable):
+        left = Argument(int)
+        right = Argument(int)
+
+    class FloatAddRhs(IntBinop):
+        right = Argument(float)
+
+    class FloatAddClip(FloatAddRhs):
+        left = Argument(float)
+        clip_lower = Argument(int, default=0)
+        clip_upper = Argument(int, default=10)
+
+    class IntAddClip(FloatAddClip, IntBinop):
+        pass
+
+    assert IntBinop.signature == TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(int))
+    ])
+    assert FloatAddRhs.signature == TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(float))
+    ])
+    assert FloatAddClip.signature == TypeSignature([
+        ('left', Argument(float)),
+        ('right', Argument(float)),
+        ('clip_lower', Argument(int, default=0)),
+        ('clip_upper', Argument(int, default=10))
+    ])
+    assert IntAddClip.signature == TypeSignature([
+        ('left', Argument(int)),
+        ('right', Argument(int)),
+        ('clip_lower', Argument(int, default=0)),
+        ('clip_upper', Argument(int, default=10))
+    ])
+
+
+def test_slots_are_inherited_and_overridable():
+    class Op(Annotable):
+        __slots__ = '_cache',  # first definition
+        arg = Argument(lambda x: x)
+
+    class StringOp(Op):
+        arg = Argument(str)  # inherit
```

Comment:
Should this be `Arg` instead of `Argument`?

Reply:
In `signature.py` it's defined as `Argument` and aliased as `Arg` during import. Do You mean renaming in signature.py?

Reply:
I see. No, leaving it here as `Argument` is fine.